File: leetcode/6.cpp
// https://leetcode.com/problems/zigzag-conversion/description/
#include <iostream>
#include <cstring>
#include <stdlib.h>
#include <vector>
#include <unordered_map>

using std::vector;
using std::cout;
using std::cin;
using std::endl;
using std::string;
using std::max;

class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1)
        {
            return s;
        }
        
        vector<string> ans(numRows);
        int j = 0;
        int state = 1;
        for (int i = 0; i < s.size(); i++)
        {
            ans[j].push_back(s[i]);
            if(state > 0){
                if(++j == numRows - 1) state*=-1;
            }else{
                if(++j == 0) state*=-1;
            }
        }
        string out;
        for (auto &&i : ans)
        {
            out.append(i);
        }
        
        return out;
    }

};


int main(){
    Solution S;
    cout << S.convert("PAYPALISHIRING",4) << endl;
    return 0;
}

File: leetcode/6.py
# https://leetcode.com/problems/zigzag-conversion/
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        n = len(s)
        ans = [[""] for _ in range(numRows)]
        j = 0
        state = 1
        for i in range(n):
            ans[j].append(s[i])
            if state > 0:
                j += 1
                if j == numRows - 1:
                    state *= -1
            else:
                j -= 1
                if not j:
                    state *= -1

        return "".join([i for m in ans for i in m])


S = Solution()
X = S.convert("AB", 3)
print(X)
